uniffi_macros/src/export.rs

```diff
@@ -160,3 +167,37 @@ pub fn rewrite_self_type(item: &mut Item) {
         visitor.visit_impl_item_mut(item);
     }
 }
+
+/// Rewrite the source input
+///
+/// This removes any uniffi attributes in the source.
+pub fn rewrite_source_input(mut item: Item) -> Item {
```

Comment:
Maybe this could be merged with https://github.com/mozilla/uniffi-rs/blob/0d58c94cbd2ef63554f3388d03d55984be76bb1f/uniffi_macros/src/lib.rs#L369-L380, but the comment scared me away from doing that.  At a minimum, the `syn::Item` needs to parse correctly before this function could remove the `[uniffi]` attributes.  I'm not sure how that interacts with `rust-analyzer`.

Reply:
Yeah, non-parseable code is exactly why those dummy macros exist. AFAIK rust-analyzer does some fixups to the code it submits to (attribute) proc-macros such that ones that output an empty token stream if the input is un-parseable still emit something meaningful, but I don't think it's possible for that sort of fixup to work for every possible partial input.

Reply:
So with this change, if the input is un-parsable we'll output it as-is, which can now include `#[uniffi(default = ..)]` attributes which would fail to compile.

I did a quick test of this by adding some invalid code to an item with default attributes and it seemed okay.  Both `nvim + rust-analyzer` and `cargo build` highlighted the syntax error and didn't complain about the attribute.

Reply:
Another test was adding the `#[uniffi(default)]` attributes in an item without the `#[uniffi::export]` attribute surround it it.  That one also seemed okay.  rust-analyzer doesn't complain about the invalid attributes.  `cargo build` fails with `cannot find attribute uniffi in this scope` which isn't super-user friendly but I don't think we can do anything to improve that.

Reply:
> So with this change, if the input is un-parsable we'll output it as-is, which can now include `#[uniffi(default = ..)]` attributes which would fail to compile.

Right, on function arguments there is no alternative like `#[uniffi::default()]` because proc-macro attributes are not supported in that position, AFAIK. I think I was actually planning to add `#[uniffi::default(arg_name = None)]` because of that.

> I did a quick test of this by adding some invalid code to an item with default attributes and it seemed okay. Both `nvim + rust-analyzer` and `cargo build` highlighted the syntax error and didn't complain about the attribute.

The potential problems are around auto-complete and assists not being able to function properly because the IDE doesn't know syntax nodes incomplete input map to. Not about the IDE reporting unexpected errors.

Reply:
Is there anything that needs to change based on this conversation?

Reply:
If you agree with my initial goal of minimizing errors when syn fails to parse the proc-macro input, yes. In that case the usage should be changed from

```rust
#[uniffi::export]
fn split(
    text: String,
    #[uniffi(default = " ")]
    sep: String,
    #[uniffi(default = None)]
    max_splits: Option<u32>,
) {
    // ...
}

#[uniffi::export]
impl Object {
    fn foo(
        #[uniffi::default(None)]
        param: Option<String>,
    ) -> Bar {
        // ...
    }
}
```

to

```rust
#[uniffi::export(default(sep = " ", max_splits = None))]
fn split(text: String, sep: String, max_splits: Option<u32>) {
    // ...
}

#[uniffi::export]
impl Object {
    #[uniffi::default(param = None)]
    fn foo(param: Option<String>) -> Bar {
        // ...
    }
}
```

Reply:
I think I was missing something the first time around, I'll change the syntax to match your proposal.  I personally think it's less intuitive, but having your editor not work is worse.

Reply:
I agree it's a bit less intuitive, but at the same time could be more readable in many cases. Also, I think "having your editor not work" is a bit of an overstatement right? AFAIU this is just about getting rid of an unnecessary / confusing error when the macro can't parse the input.

Reply:
Definitely an overstatement, but I was more thinking about your comments about autocomplete.  It feels bad when that stops working for mysterious reasons.